Fix failing open file mode is binary check in Bio/bgzf.py

terrycojones opened the pull request:
<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [ X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [ X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [ ] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->

In running the unit tests of a project of my own, I ran into the following problem (under Python 3.10):

```
self = <Bio.bgzf.BgzfWriter object at 0x19bc18d60>, filename = None, mode = 'w', fileobj = <_io.BytesIO object at 0x19be88d60>, compresslevel = 6

    def __init__(self, filename=None, mode="w", fileobj=None, compresslevel=6):
        """Initilize the class."""
        if filename and fileobj:
            raise ValueError("Supply either filename or fileobj, not both")
        if fileobj:
>           if "b" not in fileobj.mode.lower():
E           AttributeError: '_io.BytesIO' object has no attribute 'mode'

../../miniconda3/envs/310/lib/python3.10/site-packages/Bio/bgzf.py:802: AttributeError
```

I had a look at a `BytesIO` instance, and indeed there is no `mode` attribute:

```
$ python
Python 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:24:27) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import io
>>> i = io.BytesIO()
>>> dir(i)
['__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__lt__', '__ne__', '__new__', '__next__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '_checkClosed', '_checkReadable', '_checkSeekable', '_checkWritable', 'close', 'closed', 'detach', 'fileno', 'flush', 'getbuffer', 'getvalue', 'isatty', 'read', 'read1', 'readable', 'readinto', 'readinto1', 'readline', 'readlines', 'seek', 'seekable', 'tell', 'truncate', 'writable', 'write', 'writelines']
```

Based on https://stackoverflow.com/questions/44584829/how-to-determine-if-file-is-opened-in-binary-or-text-mode this pull request adds an internal `_is_binary` function to `Bio/bgzy.py` and uses it in two places.

I tried running the tests via `python setup.py build` then `python setup.py test` and got 350 errors:

```
Ran 569 tests in 102.190 seconds

FAILED (failures = 350)
```

mdehoon commented on Bio/bgzf.py:
```diff
@@ -491,6 +492,20 @@ def _load_bgzf_block(handle, text_mode=False):
+    else:
+        return "b" in fileobj.mode.lower()
+
+
```
instead you could check if `fileobj.read(0)` returns `b""`. This may be a better test, as ultimately this module relies on `fileobj.read` returning a `bytes` object.

peterjc replied:
I like the reading zero approach, and would not give it a helper function.

terrycojones replied:
OK, I changed it to use the comparison with `b""` in both places. I don't know whether this is guaranteed to work on the earliest version of Python supported by BioPython. Certainly in modern 3.x versions `"" != b""`. It might be safest to put the original test (on the `mode` attribute) back in place, and only test `read(0)` in the case of an `AttributeError`. Let me know what you prefer. Thanks for reviewing!

peterjc reviewed (approved):
Looks good to me, I would squash-and-merge this.

We require and test on at least Python 3.7, so that's fine.

peterjc commented:
Actually, looking at this afresh, ideally we'd also add a test case using BytesIO in https://github.com/biopython/biopython/blob/master/Tests/test_bgzf.py

Are you willing to add that too please?

terrycojones commented:
Yes, of course happy to add tests. I looked at `test_bgzf.py` originally, but stopped thinking about it when I tried running the tests and got those 350 errors (mentioned in original pr description). I will look again. Actually, running `pytest test_bgzf.py` in the `Tests` dir works (24 passed tests), so I will just do things that way for now.

terrycojones commented:
OK, I added four simple tests. Passing a binary-mode fileobj was not being tested, so I added tests for that too. In those cases, instead of just creating the class instances (without testing anything explicitly) I call `tell()`. I don't really consider this necessary, but some people don't like tests that don't have any asserts in them.

BTW, I put `r` before my expected error strings out of habit (I usually have trailing periods in my error messages, and need to match them with `\.`). I could remove the `r`s if you prefer them gone.

peterjc commented:
That looks great, thank you.

I missed your test issue in the original description - I think you'd found a bit of out of date instructions, sorry - see #4278

terrycojones commented:
OK, thanks for all the rapid reviewing & help re tests.
